### ISSUE TITLE
Support for closing a consumer without auto-committing offsets

### ISSUE THREAD

**Issue by braedon** [dpkp/kafka-python#932]
When I destroy a consumer I want to be able to call `close()` on it so it is removed from the consumer group (without having to wait for the session timeout). 

Calling close triggers a offset commit when using auto-commit, which assumes you have finished processing all messages you have retrieved via `poll()` or the iterator. However, if you are closing the consumer due to an error or interrupt you may not have finished processing the last message(s).

There doesn't seem to be a way to close the consumer without committing offsets when using auto-commit.

**Comment by Drizzt1991** [comment]
Hi there, I'm pretty sure it is the intended behaviour. If you enable `autocommit` you take responsibility to process all messages taken from `poll()` call. If you want to be able to skip commiting messages - just set to autocommit=False and do it yourself:
```python
consumer = KafkaConsumer(..., enable_auto_commit=False)
while not _cancelled:
    data = consumer.poll()
    # process data ...
    consumer.commit()
```
That way you will not commit messages if processing failed.

Also adding a skip option will not fix your problem, as autocommit's API does not guaranty it will not commit offsets before you call `close`. Note the `auto_commit_interval_ms ` option.

**Comment by jeffwidman** [comment]
Agreed, this should be closed as "works as designed".

**Comment by braedon** [comment]
While I agree that it is working as intended for the "success" case, what I'm looking for is the ability to more elegantly handle failure, i.e. new functionality. Given that all the consumers I have written (for production at least) are intended to run indefinitely, closing on failure is more common than success.

While I could get around this by committing myself, I'd prefer to use autocommit. There are significant benefits to committing on a regular interval, and I'd prefer not to have to build that functionality again in all my consumers myself given that it's already there.

My understanding of the API's contract is that I process all messages I receive before calling poll() again. As the new consumer is all single threaded, there is no opportunity to commit offsets after poll has returned messages, no matter how much time has passed, until control is passed back. This avoids the issue you describe, so long as you don't try and close().

Note that the issue you describe did exist on the old consumer API, where autocommitting was done in a background thread. This made it essentially unusable in most cases, as it could easily cause messages to be skipped.

**Comment by braedon** [comment]
(oops, phone slipped and pressed 'close issue' rather than 'comment'...)

**Comment by dpkp** [comment]
I put up a small PR to address this. I agree that it diverges from the java client implementation and that autocommit-on-close is generally the intended use case. But hey we're pythonistas, and adding an optional kwarg doesn't seem too bad. Thoughts?

**Comment by braedon** [comment]
Thanks @dpkp, looks like it would solve my use case!

**Comment by dpkp** [comment]
Fixed in #1031
